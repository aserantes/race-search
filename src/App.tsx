import React, { FC } from "react";
import CssBaseline from "@material-ui/core/CssBaseline";
import Container from "@material-ui/core/Container";

import { GameSchedule, Menu, GameData } from "./features";

export const App: FC = () => {
  return (
    <>
      <CssBaseline />
      <Container maxWidth="md" disableGutters>
        <Menu />
        <GameSchedule />
        <GameData />
      </Container>
    </>
  );
};
