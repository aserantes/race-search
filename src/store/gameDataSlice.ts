import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import { RootState } from ".";

interface RaceStart {
  number: number;
  driver: {
    firstName: string;
    lastName: string;
  };
  horse: {
    name: string;
    trainer: {
      firstName: string;
      lastName: string;
    };
    pedigree: {
      father: {
        name: string;
      };
    };
  };
}

interface GameRace {
  id: string;
  date: string;
  name: string;
  number: number;
  scheduledStartTime: string;
  starts: RaceStart[];
}

export interface GameDataResults {
  id: string;
  races: GameRace[];
  status: string;
}

interface GameDataState {
  gameDataResults: GameDataResults | null;
  fetchState: "idle" | "pending" | "fulfilled" | "rejected";
  errorMessage: string;
}

const initialState: GameDataState = {
  gameDataResults: null,
  fetchState: "idle",
  errorMessage: "",
};

export const fetchGameById = createAsyncThunk(
  "gameData/fetchGameById",
  async (gameId: string, { rejectWithValue }) => {
    const response = await fetch(
      `https://www.atg.se/services/racinginfo/v1/api/games/${gameId}`
    );
    if (!response.ok) {
      return rejectWithValue(`${response.status} (${response.statusText})`);
    } else {
      return (await response.json()) as GameDataResults;
    }
  }
);

const gameDataSlice = createSlice({
  name: "gameData",
  initialState,
  reducers: {
    resetGameData: (state) => {
      state.gameDataResults = null;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchGameById.fulfilled, (state, action) => {
        state.fetchState = "fulfilled";
        state.gameDataResults = action.payload;
      })
      .addCase(fetchGameById.rejected, (state, action) => {
        state.fetchState = "rejected";
        state.errorMessage =
          (action.payload as string) ||
          action.error?.message ||
          "Unknown Error";
      })
      .addCase(fetchGameById.pending, (state, action) => {
        state.fetchState = "pending";
      });
  },
});

const { actions, reducer } = gameDataSlice;

// actions
export const { resetGameData } = actions;

// selectors
export const selectGameData = (state: RootState) =>
  state.gameData.gameDataResults;
export const selectGameDataFetchState = (state: RootState) =>
  state.gameData.fetchState;
export const selectGameDataFetchError = (state: RootState) =>
  state.gameData.errorMessage;

// reducer
export default reducer;
