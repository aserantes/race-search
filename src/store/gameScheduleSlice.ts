import { createSlice, createAsyncThunk, PayloadAction } from "@reduxjs/toolkit";
import { RootState } from "./";

export interface GameInfo {
  id: string;
  startTime: string;
}

interface GameScheduleResults {
  betType: string;
  upcoming: GameInfo[];
  results: GameInfo[];
}

interface GameScheduleState {
  gameScheduleResults: GameScheduleResults | null;
  fetchState: "idle" | "pending" | "fulfilled" | "rejected";
  gameId: string;
  errorMessage: string;
}

const initialState: GameScheduleState = {
  gameScheduleResults: null,
  fetchState: "idle",
  gameId: "",
  errorMessage: "",
};

export const fetchGamesByType = createAsyncThunk(
  "gameSchedule/fetchGamesByType",
  async (gameType: string, { rejectWithValue }) => {
    const response = await fetch(
      `https://www.atg.se/services/racinginfo/v1/api/products/${gameType}`
    );
    if (!response.ok) {
      return rejectWithValue(`${response.status} (${response.statusText})`);
    } else {
      return (await response.json()) as GameScheduleResults;
    }
  }
);

const gameScheduleSlice = createSlice({
  name: "gameSchedule",
  initialState,
  reducers: {
    setGameId: (state, action: PayloadAction<string>) => {
      state.gameId = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchGamesByType.fulfilled, (state, action) => {
        state.fetchState = "fulfilled";
        state.gameScheduleResults = action.payload;
      })
      .addCase(fetchGamesByType.rejected, (state, action) => {
        state.fetchState = "rejected";
        state.errorMessage =
          (action.payload as string) ||
          action.error?.message ||
          "Unknown Error";
      })
      .addCase(fetchGamesByType.pending, (state, action) => {
        state.fetchState = "pending";
      });
  },
});

const { actions, reducer } = gameScheduleSlice;

// actions
export const { setGameId } = actions;

// selectors
export const selectGameSchedule = (state: RootState) =>
  state.gameSchedule.gameScheduleResults;
export const selectGameScheduleFetchState = (state: RootState) =>
  state.gameSchedule.fetchState;
export const selectUpcoming = (state: RootState) =>
  state.gameSchedule.gameScheduleResults?.upcoming;
export const selectResults = (state: RootState) =>
  state.gameSchedule.gameScheduleResults?.results;
export const selectGameScheduleFetchError = (state: RootState) =>
  state.gameSchedule.errorMessage;
export const selectGameId = (state: RootState) => state.gameSchedule.gameId;

// reducer
export default reducer;
