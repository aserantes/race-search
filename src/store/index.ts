import { configureStore, ThunkAction, Action } from "@reduxjs/toolkit";
import searchInputReducer from "./searchInputSlice";
import gameScheduleReducer from "./gameScheduleSlice";
import gameDataReducer from "./gameDataSlice";

export const store = configureStore({
  reducer: {
    searchInput: searchInputReducer,
    gameSchedule: gameScheduleReducer,
    gameData: gameDataReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export { setGameType, selectGameType } from "./searchInputSlice";

export {
  fetchGamesByType,
  setGameId,
  selectGameSchedule,
  selectGameScheduleFetchState,
  selectUpcoming,
  selectResults,
  selectGameScheduleFetchError,
  selectGameId,
} from "./gameScheduleSlice";

export type { GameInfo } from "./gameScheduleSlice";

export {
  fetchGameById,
  resetGameData,
  selectGameDataFetchState,
  selectGameData,
  selectGameDataFetchError,
} from "./gameDataSlice";

export type { GameDataResults } from "./gameDataSlice";
