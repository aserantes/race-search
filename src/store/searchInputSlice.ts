import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { RootState } from "./";

interface SearchInputState {
  gameType: string;
}

const initialState: SearchInputState = {
  gameType: "",
};

const searchInputSlice = createSlice({
  name: "searchInput",
  initialState,
  reducers: {
    setGameType: (state, action: PayloadAction<string>) => {
      state.gameType = action.payload;
    },
  },
});

const { actions, reducer } = searchInputSlice;

// actions
export const { setGameType } = actions;

// selectors
export const selectGameType = (state: RootState) => state.searchInput.gameType;

// reducer
export default reducer;
