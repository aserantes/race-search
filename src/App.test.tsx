import React from "react";
import { render, cleanup } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "./store";
import App from "./App";

afterEach(cleanup);

it("should match a snapshot", () => {
  const { asFragment } = render(
    <Provider store={store}>
      <App />
    </Provider>
  );

  expect(
    asFragment(
      <Provider store={store}>
        <App />
      </Provider>
    )
  ).toMatchSnapshot();
});
