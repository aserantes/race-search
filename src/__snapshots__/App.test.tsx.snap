// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match a snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-disableGutters MuiContainer-maxWidthMd"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <h5
          class="MuiTypography-root makeStyles-title-3 MuiTypography-h5"
        >
          Race Search
        </h5>
      </div>
    </header>
    <div>
      <div>
        SearchInput
      </div>
      <input
        type="text"
        value=""
      />
      <div>
        real-time:  debounced: 
      </div>
    </div>
    <div>
      <div>
        GameSchedule
      </div>
    </div>
  </div>
</DocumentFragment>
`;
